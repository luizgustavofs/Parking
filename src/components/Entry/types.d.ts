export type EntryState = 'default' | 'error' | 'loading' | 'success';

export type EntryProps = {
  plate: string;
  setPlate: (newPlate: string) => void;
};
