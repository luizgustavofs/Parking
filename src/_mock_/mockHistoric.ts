export const MOCK_HISTORIC_EMPTY = [];

export const MOCK_HISTORIC_WITHOUT_LEFT = [
  {
    time: '12 days 5 hours',
    paid: true,
    left: false,
    plate: 'AAA-1111',
    reservation: '607ecf',
  },
  {
    time: '37 seconds',
    paid: true,
    left: true,
    plate: 'AAA-1111',
    reservation: '6046d8',
  },
  {
    time: '20 minutes',
    paid: true,
    left: true,
    plate: 'AAA-1111',
    reservation: '60451e',
  },
  {
    time: '12 minutes',
    paid: true,
    left: true,
    plate: 'AAA-1111',
    reservation: '6047d2',
  },
  {
    time: '52 seconds',
    paid: true,
    left: true,
    plate: 'AAA-1111',
    reservation: '6046d9',
  },
  {
    time: '33 seconds',
    paid: true,
    left: true,
    plate: 'AAA-1111',
    reservation: '6046d9',
  },
];

export const MOCK_HISTORIC_WITHOUT_PAYMENT = [
  {
    time: '12 days 5 hours',
    paid: false,
    left: false,
    plate: 'AAA-1111',
    reservation: '607ecf',
  },
  {
    time: '37 seconds',
    paid: true,
    left: true,
    plate: 'AAA-1111',
    reservation: '6046d8',
  },
  {
    time: '20 minutes',
    paid: true,
    left: true,
    plate: 'AAA-1111',
    reservation: '60451e',
  },
  {
    time: '12 minutes',
    paid: true,
    left: true,
    plate: 'AAA-1111',
    reservation: '6047d2',
  },
  {
    time: '52 seconds',
    paid: true,
    left: true,
    plate: 'AAA-1111',
    reservation: '6046d9',
  },
  {
    time: '33 seconds',
    paid: true,
    left: true,
    plate: 'AAA-1111',
    reservation: '6046d9',
  },
];

export const MOCK_HISTORIC_COMPLETE = [
  {
    time: '12 days 5 hours',
    paid: true,
    left: true,
    plate: 'AAA-1111',
    reservation: '607ecf',
  },
  {
    time: '37 seconds',
    paid: true,
    left: true,
    plate: 'AAA-1111',
    reservation: '6046d8',
  },
  {
    time: '20 minutes',
    paid: true,
    left: true,
    plate: 'AAA-1111',
    reservation: '60451e',
  },
  {
    time: '12 minutes',
    paid: true,
    left: true,
    plate: 'AAA-1111',
    reservation: '6047d2',
  },
  {
    time: '52 seconds',
    paid: true,
    left: true,
    plate: 'AAA-1111',
    reservation: '6046d9',
  },
  {
    time: '33 seconds',
    paid: true,
    left: true,
    plate: 'AAA-1111',
    reservation: '6046d9',
  },
];
