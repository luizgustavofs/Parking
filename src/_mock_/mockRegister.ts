export const MOCK_REGISTER_OK = {
  reservation: '62733f',
  plate: 'AAA-0044',
  entered_at: '2022-05-05T03:07:50.710+00:00',
};
