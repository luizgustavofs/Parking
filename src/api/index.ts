import axios from 'axios';

export const api = axios.create({
  baseURL: 'https://parking-lot-to-pfz.herokuapp.com',
  timeout: 5000,
  headers: {},
});
