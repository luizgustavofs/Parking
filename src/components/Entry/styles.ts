import styled from 'styled-components';

export const Container = styled.div`
  min-height: 300px;
  padding: 16px;
  width: 100%;

  align-items: center;
  flex-direction: column;
  justify-content: center;

  background-color: white;

  > img {
    width: 130px;
    height: 130px;
  }

  @media (min-width: 1024px) {
    width: 400px;
  }
`;
