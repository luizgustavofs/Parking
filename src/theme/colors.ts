export const colors = {
  default: '#000000',
  primary: '#4DD0E1',
  secondary: '#F2F2F2',
  tertiary: '#28DD91',
  quaternary: '#A769B2',
  error: '#FF0068',
  baseColor: '#FFFBE6',
  neutral: {
    _00: '#ffffff',
    _70: '#DADADA',
    _80: '#9B9B9B',
  },
};
