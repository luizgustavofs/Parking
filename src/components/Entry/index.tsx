/* eslint-disable @next/next/no-img-element */
import React, { useState } from 'react';

import { useTranslation } from 'react-i18next';
import ReactLoading from 'react-loading';
import { toast } from 'react-toastify';
import { useTheme } from 'styled-components';

import {
  getVehicleHistory,
  registerVehicleEntry,
} from '../../services/parking';
import Button from '../Button';
import Input from '../Input';
import { Container } from './styles';
import { EntryProps, EntryState } from './types';

const Entry: React.FC<EntryProps> = ({ plate, setPlate }) => {
  const { colors } = useTheme();
  const [entryState, setEntryState] = useState<EntryState>('default');
  const { t } = useTranslation();

  const checkCanRegister = async () => {
    setEntryState('loading');
    const historic = await getVehicleHistory(plate).catch(error => {
      toast.error(t('ERROR.API_ERROR'));
      console.log(error);
      setEntryState('default');
    });

    if (!historic) return;

    if (historic?.length === 0) {
      return registerPlate();
    } else {
      const notCompletedRegister = historic.find(
        element => element.left === false || element.paid === false,
      );

      if (!notCompletedRegister) {
        registerPlate();
      } else {
        toast.error(t('ENTRY.ERROR-NEW-VEHICLE'));
        setEntryState('default');
      }
    }
  };

  const registerPlate = async () => {
    setEntryState('loading');
    const register = await registerVehicleEntry(plate).catch(error => {
      toast.error(t('ERROR.API_ERROR'));
      console.log(error);
      setEntryState('default');
    });

    if (!register) return;

    if (register.entered_at) {
      setEntryState('success');
      toast.success(t('ENTRY.SUCCESS'));
      setTimeout(() => setEntryState('default'), 2000);
    } else {
      setEntryState('default');
      toast.error(t('ENTRY.ERROR-NEW-VEHICLE'));
    }
  };

  return (
    <Container>
      {entryState === 'loading' && (
        <>
          <ReactLoading
            type={'spokes'}
            color={colors.primary}
            height={130}
            width={130}
          />
          <h4>{t('ENTRY.LOADING')}</h4>
        </>
      )}
      {entryState === 'success' && (
        <>
          <img src={'/round-done-button.svg'} alt="Ícone de sucesso" />
          <h4>{t('ENTRY.REGISTER')}</h4>
        </>
      )}
      {(entryState === 'default' || entryState === 'error') && (
        <>
          <Input
            testID="plate-input"
            label={t('ENTRY.INPUT.LABEL')}
            placeholderText={t('ENTRY.INPUT.PLACEHOLDER')}
            value={plate}
            onChange={value => {
              setPlate(value);
              setEntryState('default');
            }}
            color={entryState === 'error' ? colors.error : colors.default}
          />
          <Button
            testID="button-confirm"
            primary
            disabled={plate === '' || plate.includes('_')}
            onClick={checkCanRegister}>
            {t('ENTRY.CONFIRM')}
          </Button>
        </>
      )}
    </Container>
  );
};

export default Entry;
