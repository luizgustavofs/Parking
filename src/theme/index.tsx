import { ReactNode } from 'react';

import { ThemeProvider } from 'styled-components';

import { colors } from './colors';
import { fontSizes } from './fontSizes';
import { spaces } from './spaces';

export const theme = {
  colors,
  spaces,
  fontSizes,
};

const ProjectThemeProvider: React.FC<{ children: ReactNode }> = ({
  children,
  ...props
}) => {
  return (
    <ThemeProvider theme={theme} {...props}>
      {children}
    </ThemeProvider>
  );
};

export { ProjectThemeProvider };
