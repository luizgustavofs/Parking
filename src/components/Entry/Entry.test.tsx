import { fireEvent, waitFor } from '@testing-library/dom';
import { render } from '@testing-library/react';
import AxiosMockAdapter from 'axios-mock-adapter';
import { act, create } from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';

import Entry from '.';
import {
  MOCK_HISTORIC_COMPLETE,
  MOCK_HISTORIC_EMPTY,
  MOCK_HISTORIC_WITHOUT_LEFT,
  MOCK_HISTORIC_WITHOUT_PAYMENT,
} from '../../_mock_/mockHistoric';
import { MOCK_REGISTER_OK } from '../../_mock_/mockRegister';
import { api } from '../../api';
import { mockToastError, mockToastSuccess } from '../../tests/setupTests';
import { theme } from '../../theme';

describe('Entry', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly with variants plate and setPlate', () => {
    const { toJSON } = create(
      <ThemeProvider theme={theme}>
        <Entry plate="AAA-0000" setPlate={jest.fn} />
      </ThemeProvider>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('should button be disable if not have plate', () => {
    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <Entry plate="" setPlate={jest.fn} />
      </ThemeProvider>,
    );

    const button = getByTestId('button-confirm');

    expect(button).toHaveProperty('disabled', true);
  });

  it('should add plate when button is clicked', () => {
    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <Entry plate="AAA-0000" setPlate={jest.fn} />
      </ThemeProvider>,
    );

    const button = getByTestId('button-confirm');
    fireEvent.click(button);
    expect(button).toHaveBeenCalled;
  });

  it('should add plate and show success toast when everything is ok', async () => {
    var mock = new AxiosMockAdapter(api);

    mock.onGet('/parking/AAA-1112').reply(200, MOCK_HISTORIC_COMPLETE);
    mock.onPost('/parking').reply(200, MOCK_REGISTER_OK);

    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <Entry plate="AAA-1112" setPlate={jest.fn} />
      </ThemeProvider>,
    );

    const button = getByTestId('button-confirm');
    await fireEvent.click(button);

    await waitFor(() => expect(mockToastSuccess).toBeCalled());
  });

  it('should not add plate and show error toast when plate still not left', async () => {
    var mock = new AxiosMockAdapter(api);

    mock.onGet('/parking/AAA-1112').reply(200, MOCK_HISTORIC_WITHOUT_LEFT);
    mock.onPost('/parking').reply(200, MOCK_REGISTER_OK);

    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <Entry plate="AAA-1112" setPlate={jest.fn} />
      </ThemeProvider>,
    );

    const button = getByTestId('button-confirm');
    await fireEvent.click(button);

    await waitFor(() => expect(mockToastError).toBeCalled());
  });

  it('should not add plate and show error toast when plate still not paid', async () => {
    var mock = new AxiosMockAdapter(api);

    mock.onGet('/parking/AAA-1112').reply(200, MOCK_HISTORIC_WITHOUT_PAYMENT);
    mock.onPost('/parking').reply(200, MOCK_REGISTER_OK);

    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <Entry plate="AAA-1112" setPlate={jest.fn} />
      </ThemeProvider>,
    );

    const button = getByTestId('button-confirm');
    await act(async () => {
      await fireEvent.click(button);
    });

    await waitFor(() => expect(mockToastError).toBeCalled());
  });

  it('should add plate and show error toast when plate still not paid', async () => {
    var mock = new AxiosMockAdapter(api);
    jest.useFakeTimers();

    mock.onGet('/parking/AAA-1112').reply(200, MOCK_HISTORIC_EMPTY);
    mock.onPost('/parking').reply(200, MOCK_REGISTER_OK);

    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <Entry plate="AAA-1112" setPlate={jest.fn} />
      </ThemeProvider>,
    );

    const button = getByTestId('button-confirm');
    await act(async () => {
      await fireEvent.click(button);
    });

    await waitFor(() => expect(mockToastSuccess).toBeCalled());
  });

  it('should add plate and show error toast when plate still not paid', async () => {
    var mock = new AxiosMockAdapter(api);
    jest.useFakeTimers();

    mock.onGet('/parking/AAA-1112').reply(200, MOCK_HISTORIC_EMPTY);
    mock.onPost('/parking').reply(200, {});

    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <Entry plate="AAA-1112" setPlate={jest.fn} />
      </ThemeProvider>,
    );

    const button = getByTestId('button-confirm');
    await act(async () => {
      await fireEvent.click(button);
    });

    await waitFor(() => expect(mockToastError).toBeCalled());
  });

  it('should change de value input if change plate', () => {
    var mock = new AxiosMockAdapter(api);

    mock.onGet('/parking/AAA-0000').reply(200, MOCK_HISTORIC_COMPLETE);
    mock.onPost('/parking/AAA-0000').reply(200, {});

    const changePlateMock = jest.fn();

    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <Entry plate="AAA-0000" setPlate={changePlateMock} />
      </ThemeProvider>,
    );

    const input = getByTestId('plate-input');
    fireEvent.change(input, { target: { value: 'NEW-0000' } });

    expect(changePlateMock).toHaveBeenCalledWith('NEW-0000');
  });
});
