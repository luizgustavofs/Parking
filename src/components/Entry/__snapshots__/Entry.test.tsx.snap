// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Entry should render correctly with variants plate and setPlate 1`] = `
<div
  className="sc-dkzDqf bznBgJ"
>
  <div
    className="sc-gsnTZi bYaekj"
    color="#000000"
  >
    <h3>
      ENTRY.INPUT.LABEL
    </h3>
    <input
      color="#000000"
      data-testid="plate-input"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onMouseDown={[Function]}
      onPaste={[Function]}
      placeholder="ENTRY.INPUT.PLACEHOLDER"
      value="AAA-0000"
    />
  </div>
  <button
    className="sc-bczRLJ DNMyx"
    color="#28DD91"
    data-testid="button-confirm"
    disabled={false}
    onClick={[Function]}
  >
    <h4>
      ENTRY.CONFIRM
    </h4>
  </button>
</div>
`;
